src/qpx_transport.cpp

```diff
@@ -0,0 +1,1170 @@
+                     const unsigned char *sense) {
+  int ret = -1;
+
+  if (ioret == kIOReturnSuccess)
+    ret = 0;
+  else if (ioret == kIOReturnNoDevice)
+    errno = ENXIO;
+  else if (ioret == kIOReturnNoMemory)
+    errno = ENOMEM;
+  else if (ioret == kIOReturnExclusiveAccess)
+    errno = EBUSY;
+  else
+    errno = EIO;
+
+  if (ret)
+    return ret;
+
+  if (stat == kSCSITaskStatus_CHECK_CONDITION) {
+    ret = ERRCODE(sense);
+    if (ret == 0)
+      errno = EIO, ret = -1;
+    else
+      CREAM_ON_ERRNO(sense);
+  } else if (stat != kSCSITaskStatus_GOOD)
+    errno = EIO, ret = -1;
+
+  return ret;
+}
+
+Scsi_Command::Scsi_Command() {
+  scsiob = IO_OBJECT_NULL, plugin = NULL, mmcdif = NULL, taskif = NULL;
+  autoclose = 1;
+  filename = NULL;
+}
+
+Scsi_Command::Scsi_Command(void *f) {
+  taskif = (SCSITaskDeviceInterface **)f, autoclose = 0;
+  filename = NULL;
+}
+
+Scsi_Command::~Scsi_Command() {
+  if (autoclose) {
+    if (taskif)
+      (*taskif)->ReleaseExclusiveAccess(taskif), (*taskif)->Release(taskif),
+          taskif = NULL;
+    if (mmcdif)
+      (*mmcdif)->Release(mmcdif), mmcdif = NULL;
+    if (plugin)
+      IODestroyPlugInInterface(plugin), plugin = NULL;
+    if (scsiob)
+      IOObjectRelease(scsiob), scsiob = IO_OBJECT_NULL;
+  }
+  if (filename)
+    free(filename), filename = NULL;
+}
+
+int Scsi_Command::associate(const char *file, const struct stat *ref) {
+  struct stat sb;
+  io_object_t scsiob = IO_OBJECT_NULL, parent;
+  CFMutableDictionaryRef match, bsddev;
+  CFNumberRef num;
+  int i;
+
+  if (ref)
+    sb = *ref;
+  else if (stat(file, &sb))
+    return 0;
+
+  if (!(S_ISBLK(sb.st_mode) || S_ISCHR(sb.st_mode)))
+    return !(errno = ENOTBLK);
+
+  if ((match = CFDictionaryCreateMutable(
+           kCFAllocatorDefault, 0, &kCFTypeDictionaryKeyCallBacks,
+           &kCFTypeDictionaryValueCallBacks)) == NULL)
+    return !(errno = ENOMEM);
+  if ((bsddev = CFDictionaryCreateMutable(
+           kCFAllocatorDefault, 0, &kCFTypeDictionaryKeyCallBacks,
+           &kCFTypeDictionaryValueCallBacks)) == NULL)
+    return CFRelease(match), !(errno = ENOMEM);
+
+  i = major(sb.st_rdev);
+  num = CFNumberCreate(kCFAllocatorDefault, kCFNumberIntType, &i);
+  CFDictionarySetValue(bsddev, CFSTR("BSD Major"), num);
+  CFRelease(num);
+
+  i = minor(sb.st_rdev);
+  num = CFNumberCreate(kCFAllocatorDefault, kCFNumberIntType, &i);
+  CFDictionarySetValue(bsddev, CFSTR("BSD Minor"), num);
+  CFRelease(num);
+
+  CFDictionarySetValue(match, CFSTR(kIOPropertyMatchKey), bsddev);
+  CFRelease(bsddev);
+
+  if ((scsiob = IOServiceGetMatchingService(kIOMasterPortDefault, match)) ==
+      IO_OBJECT_NULL)
+    return !(errno = ENXIO);
+
+  // traverse up to "SCSITaskAuthoringDevice"
+  kern_return_t kret;
+  while ((kret = IORegistryEntryGetParentEntry(scsiob, kIOServicePlane,
+                                               &parent)) == kIOReturnSuccess) {
+    CFStringRef uclient;
+    const char *s;
```

Comment:
Codacy found an issue: [The scope of the variable 's' can be reduced.](https://app.codacy.com/manual/jeremiahwilson777/qpxtool/pullRequest?prid=6191247)